core/src/main/java/com/linecorp/armeria/internal/common/stream/PrependingPublisher.java

```diff
@@ -0,0 +1,135 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public final class PrependingPublisher<T> implements Publisher<T> {
+
+    private final T first;
+    private final Publisher<? extends T> rest;
+
+    public PrependingPublisher(T first, Publisher<? extends T> rest) {
+        this.first = first;
+        this.rest = rest;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        final RestSubscriber restSubscriber = new RestSubscriber(subscriber);
+        rest.subscribe(restSubscriber);
+    }
+
+    final class RestSubscriber implements Subscriber<T>, Subscription {
+
+        private Subscriber<? super T> downstream;
+        private final AtomicLong demand = new AtomicLong();
+        @Nullable
+        private volatile Subscription upstream;
+        @Nullable
+        private volatile Throwable upstreamCause;
+        private volatile boolean upstreamCompleted;
+        private volatile boolean completed;
+        private volatile boolean firstSent;
+
+        RestSubscriber(Subscriber<? super T> downstream) {
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            this.upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T t) {
+            demand.decrementAndGet();
+            downstream.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            // delay onError until the first piece is sent
+            if (!firstSent) {
+                upstreamCause = t;
+            } else {
+                downstream.onError(t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            // delay onComplete until the first piece is sent
+            if (!firstSent) {
+                upstreamCompleted = true;
+            } else {
+                downstream.onComplete();
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (completed) {
+                return;
+            }
+            if (demand.getAndAdd(n) > 0) {
+                return;
+            }
```

Comment:
Don't we need to pass the request signal to upstream if the previous `demand` is 0?

Reply:
If previous `demand` is 0 then after `getAndAdd`, this should be able to execute https://github.com/line/armeria/blob/afd7938b14681af8941a1cd450d25b9904372d56/core/src/main/java/com/linecorp/armeria/internal/common/stream/PrependingPublisher.java#L113

I'm not sure I follow you here 😅